juniper_codegen/src/common/field/arg.rs

```diff
@@ -351,14 +351,69 @@ impl OnMethod {
         match self {
             Self::Regular(arg) => {
                 let (name, ty) = (&arg.name, &arg.ty);
-                let err_text = format!(
-                    "Internal error: missing argument `{}` - validation must have failed",
-                    &name,
-                );
+                let err_text = format!("Missing argument `{}`: {{}}", &name);
                 quote! {
                     args.get::<#ty>(#name)
-                        .or_else(::juniper::FromInputValue::<#scalar>::from_implicit_null)
-                        .expect(#err_text)
+                        .and_then(|opt| {
+                            opt.map_or_else(
+                                || {
+                                    <#ty as ::juniper::FromInputValue<#scalar>>::
+                                        from_implicit_null()
+                                        .map_err(|e| {
+                                            let e = ::juniper::IntoFieldError::into_field_error(e);
+                                            ::juniper::FieldError::new(
+                                                format!(#err_text, e.message()),
+                                                e.extensions().clone(),
+                                            )
+                                        })
+                                },
+                                Ok,
+                            )
+                        })?
```

Comment:
@ilslv  I don't follow the logic here.

It was:
1. `args.get::<#ty>(#name)` returns `Option<T>`.
2. If it's `None`, then `::juniper::FromInputValue::<#scalar>::from_implicit_null` returns `Option<T>`.
3. If it's `None` we panic.

Now it is:
1.  `args.get::<#ty>(#name)` returns `Result<Option<T>, FieldError>`.
2. If it's `Ok`, we check the inner `opt`.
3. If it `Some` we map it to `Some(Ok(v))` and so have `Result<Option<Result<T>, FieldError>, FieldError>`.

Why?

It seems the more natural way is:
1. Unpack the `Result` early: `args.get::<#ty>(#name)?`
2. `.transpose()` the inner error in `Option` and return it with `?`.

Reply:
Oh... understood. `.map_or_else` doesn't map an inner `Option` value, but rather maps the whole `Option` into something.

`.ok_or_else` will do better here.

Reply:
Meh... `.ok_or_else()` cannot map absense to `Result::Ok`.